pkg/pillar/cmd/nim/controllerdns.go

```diff
@@ -70,28 +81,191 @@ func (n *nim) queryControllerDNS() {
+	sourceUDPAddr := net.UDPAddr{IP: srcIP}
+	dialer := net.Dialer{LocalAddr: &sourceUDPAddr}
+	dnsClient := dns.Client{Dialer: &dialer}
+	msg := dns.Msg{}
+	if domain[len(domain)-1] != '.' {
+		domain = domain + "."
+	}
+	msg.SetQuestion(domain, dns.TypeA)
+	dnsClient.Timeout = time.Duration(dnsTimeout)
+	reply, _, err := dnsClient.Exchange(&msg, net.JoinHostPort(dnsServerIP.String(), "53"))
+	if err != nil {
+		n.Log.Tracef("dns exchange failed: %v", err)
+		return nil
+	}
+	if len(reply.Answer) > 0 {
+		for _, answer := range reply.Answer {
+			if aRecord, ok := answer.(*dns.A); ok {
+				response = append(response, dnsResponse{
+					ip:  aRecord.A,
+					ttl: aRecord.Header().Ttl,
+				})
+			}
+		}
+	}
+	return response
+}
+
+func (n *nim) resolveWithPorts(domain string) []dnsResponse {
+	return n.resolveWithPortsLambda(domain, n.resolveWithSrcIP)
+}
+
+func (n *nim) resolveWithPortsLambda(domain string, resolve func(string, net.IP, net.IP) []dnsResponse) []dnsResponse {
+	quit := make(chan struct{})
+	work := make(chan struct{}, dnsMaxParallelRequests)
+	resolvedIPsChan := make(chan []dnsResponse)
+	countDNSRequests := 0
+	var wg sync.WaitGroup
+
+	for _, port := range n.dpcManager.GetDNS().Ports {
+		if port.Cost > 0 {
+			continue
+		}
+
+		var srcIPs []net.IP
+		for _, addrInfo := range port.AddrInfoList {
+			srcIPs = append(srcIPs, addrInfo.Addr)
+		}
+
+		dnsInfo, ok := n.dnsInfo(port)
+		if !ok {
+			continue
+		}
+
+		for _, dnsIP := range dnsInfo.DNSServers {
+			for _, srcIP := range srcIPs {
+				wg.Add(1)
+				dnsIPCopy := make(net.IP, len(dnsIP))
+				copy(dnsIPCopy, dnsIP)
+				srcIPCopy := make(net.IP, len(srcIP))
+				copy(srcIPCopy, srcIP)
+				countDNSRequests++
+				go func(dnsIP, srcIP net.IP) {
+					select {
+					case work <- struct{}{}:
+						// if writable, means less than dnsMaxParallelRequests goroutines are currently running
+					}
+					select {
+					case <-quit:
+						// will return in case the quit chan has been closed, meaning another dns server already resolved the IP
+						return
+					default:
+					}
+					response := resolve(domain, dnsIP, srcIP)
+					if response != nil {
+						resolvedIPsChan <- response
+					}
+					<-work
+					wg.Done()
+				}(dnsIPCopy, srcIPCopy)
+			}
+		}
+	}
+
+	wgChan := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(wgChan)
+	}()
+
+	select {
+	case <-wgChan:
+		var responses []dnsResponse
+		if countDNSRequests == 0 {
+			// fallback in case no resolver is configured
+			ips, err := net.LookupIP(domain)
+			if err != nil {
+				n.Log.Warnf("fallback resolver failed: %+v", err)
+			}
+			for _, ip := range ips {
+				responses = append(responses, dnsResponse{
+					ip:  ip,
+					ttl: uint32(maxTTLSec),
+				})
+			}
+		}
+		return responses
+	case ip := <-resolvedIPsChan:
+		close(quit)
+		return ip
+	}
+}
+
+func (n *nim) dnsInfo(port types.NetworkPortStatus) (netmonitor.DNSInfo, bool) {
+	ifIndex, exist, err := n.networkMonitor.GetInterfaceIndex(port.IfName)
```

Comment:
I think this function is unnecessary - DNSInfo is part of DNS and it is updated as needed by DPCManager:  https://github.com/lf-edge/eve/blob/master/pkg/pillar/dpcmanager/dns.go#L251-L271

Reply:
removed it

Reply:
I still see `dnsInfo` in the diff, although it is not called anymore.

Reply:
Now it is really gone ...